### PR TITLE
Update refinement

### PR DESCRIPTION
**Pull request by alexfikl**
Follow up from #70 with some discussion in alexfikl/meshmode#3.

This PR mostly consolidates the refinement of elements backed by modepy shapes (from inducer/modepy#16). The actual algorithm is not changed (at least in spirit) from the existing one.

Main benefit is that it can now refine cubes, not just simplices!

### CODE REVIEW

File: meshmode/mesh/refinement/tesselate.py [hunk 1]
```diff
@@ -23,105 +24,248 @@
...
-        return tesselatetri()
-    elif dim == 3:
-        return tesselatetet()
-    else:
-        raise ValueError("cannot tesselate %d-simplex" % dim)
+    return tuple(midpoint(ai, bi) for ai, bi in zip(a, b))
+
+
+def _get_ref_midpoints(shape, ref_vertices):
+    r"""The reference element is considered to be, e.g. for a 2 simplex::
```
Comment (inducer) [inline meshmode/mesh/refinement/tesselate.py:155]:
```suggestion
    r"""The reference element is considered to be, e.g. for a 2-simplex::
```

File: meshmode/mesh/refinement/tesselate.py [hunk 2]
```diff
@@ -23,105 +24,248 @@
...
+        A :class:`TesselationInfo` that describes the tesselation of the
+        element group.
+
+    .. attribute:: element_mapping
+
+        A mapping from the original elements to the refined child elements.
+    """
+
+    tess_info: TesselationInfo
+    element_mapping: List[List[int]]
```
Comment (inducer) [inline meshmode/mesh/refinement/tesselate.py:95]:
```suggestion
    # FIXME: This should really be a CSR data structure.
    element_mapping: List[List[int]]
```

File: meshmode/mesh/refinement/tesselate.py [hunk 3]
```diff
@@ -23,105 +24,248 @@
...
-def halve_tuple(a):
-    def halve(x):
-        d, r = divmod(x, 2)
+
+# {{{ interface
+
+@dataclass(frozen=True)
+class TesselationInfo:
+    """
+    .. attribute:: children
```
Comment (inducer) [inline meshmode/mesh/refinement/tesselate.py:46]:
```suggestion
@dataclass(frozen=True)
class ElementTesselationInfo:
    """Describes how one element is split into multiple child elements.
    
    .. attribute:: children
```

File: meshmode/mesh/refinement/tesselate.py [hunk 4]
```diff
@@ -23,105 +24,248 @@
...
+    midpoint_vertex_pairs: Optional[List[Tuple[int, int]]] = None
+
+
+@dataclass(frozen=True)
+class GroupRefinementRecord:
+    """
+    .. attribute:: tess_info
+
+        A :class:`TesselationInfo` that describes the tesselation of the
+        element group.
```
Comment (inducer) [inline meshmode/mesh/refinement/tesselate.py:87]:
```suggestion
        An instance of :class:`ElementTesselationInfo` that describes the
        tesselation of a single element into multiple child elements.
```

File: meshmode/mesh/refinement/tesselate.py [hunk 5]
```diff
@@ -23,105 +24,248 @@
...
+
+    orig_vertex_indices: Optional[np.ndarray] = None
+    midpoint_indices: Optional[np.ndarray] = None
+    midpoint_vertex_pairs: Optional[List[Tuple[int, int]]] = None
+
+
+@dataclass(frozen=True)
+class GroupRefinementRecord:
+    """
+    .. attribute:: tess_info
```
Comment (inducer) [inline meshmode/mesh/refinement/tesselate.py:84]:
```suggestion
    .. attribute:: el_tess_info
```

**Review by inducer — COMMENTED** [review]
Thanks for working on this. LGTM, just a few style things left. FWIW, I looked over my comments in https://github.com/alexfikl/meshmode/pull/3, and I very much like the direction this has taken.

File: meshmode/mesh/refinement/tesselate.py [hunk 6]
```diff
@@ -23,105 +24,248 @@
...
+
+# {{{ interface
+
+@dataclass(frozen=True)
+class TesselationInfo:
+    """
+    .. attribute:: children
+
+        A tess_info of the reference element, given here by
+        :attr:`ref_vertices`.
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/refinement/tesselate.py]: Oops! That was a sed gone wrong. Hopefully improve in da75962.

File: meshmode/mesh/refinement/tesselate.py (hunk 5 above)
Comment (alexfikl) [reply meshmode/mesh/refinement/tesselate.py]: Renamed in ce36d6a.

**Comment by inducer** [comment]
LGTM. Thanks for working on this, too!

**Pull request merged**